samples/apps/tinyRA/tinyra/tui.py

```diff
@@ -0,0 +1,1186 @@
+        pretty_print (bool, optional): Whether to pretty print the JSON. Defaults to True.
+
+    Returns:
+        str: The markdown code block representing the JSON object.
+    """
+    if pretty_print:
+        json_string = json.dumps(json_data, indent=2)
+    else:
+        json_string = json.dumps(json_data)
+
+    markdown_code_block = f"```json\n{json_string}\n```"
+    return markdown_code_block
+
+
+def message2markdown(message: Dict[str, str]) -> str:
+    """
+    Convert a message to markdown that can be displayed in the chat display.
+
+    Args:
+        message: a message
+
+    Returns:
+        A markdown string.
+    """
+    role = message["role"]
+    if role == "user":
+        display_name = APP_CONFIG.get_user_name()
+    elif role == "assistant" or role == "info":
+        display_name = "TinyRA"
+    else:
+        display_name = "."
+
+    if role == "info":
+        display_id = "\U0001F4AD" * 3
+    else:
+        display_id = message["id"]
+
+    content = message["content"]
+
+    return f"[{display_id}] {display_name}: {content}"
+
+
+class ReactiveMessage(Markdown):
+    """
+    A reactive markdown widget for displaying assistant messages.
+    """
+
+    message = reactive({"role": None, "content": None, "id": None})
+
+    class Selected(Message):
+        """Assistant message selected message."""
+
+        def __init__(self, msg_id: str) -> None:
+            self.msg_id = msg_id
+            super().__init__()
+
+    def __init__(self, id=None, role=None, content=None, **kwargs):
+        super().__init__(**kwargs)
+        self.message = {"role": role, "content": content, "id": id}
+
+    def on_mount(self) -> None:
+        self.set_interval(1, self.update_message)
+        chat_display = self.app.query_one(ChatDisplay)
+        chat_display.scroll_end()
+
+    def on_click(self) -> None:
+        self.post_message(self.Selected(self.message.get("id")))
+
+    def update_message(self):
+        message = fetch_row(self.message.get("id"))
+
+        if message is None:
+            self.remove()
+            return
+
+        self.classes = f"{self.message['role'].lower()}-message message"
+
+        self.message = message
+
+    async def watch_message(self) -> str:
+        await self.update(message2markdown(self.message))
+
+
+def message_display_handler(message: Dict[str, str]):
+    """
+    Given a message, return a widget for displaying the message.
+    If the message is from the user, return a markdown widget.
+    If the message is from the assistant, return a reactive markdown widget.
+
+    Args:
+        message: a message
+
+    Returns:
+        A markdown widget or a reactive markdown widget.
+    """
+    role = message["role"]
+    id = message["id"]
+    content = message["content"]
+    message_widget = ReactiveMessage(id=id, role=role, content=content, classes=f"{role.lower()}-message message")
+    return message_widget
+
+
+class DirectoryTreeContainer(ScrollableContainer):
```

Comment:
A `DirectoryTree` already is a scrollable container. Rather than a DT inside a container, could you extend the `DirectoryTree`?

It can cause a bit of confusion if you have a scrollable thing inside another scrollable thing. It's not always clear what widget the scrollbar is scrolling...